tee_supplicant/src/tee_supplicant.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/tee.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <optee_msg_supplicant.h>
+#include <tee_client_api.h>
+#include <teec_ta_load.h>
+
+LOG_MODULE_REGISTER(tee_supplicant);
+
+#define TEE_SUPP_THREAD_PRIO	7
+
+#define TEE_REQ_PARAM_MAX	5
+
+static struct k_thread main_thread;
+static K_THREAD_STACK_DEFINE(main_stack, 8192);
+
+static K_MUTEX_DEFINE(shm_mutex);
+
+#define MEMREF_SHM_ID(p)	((p)->c)
+#define MEMREF_SHM_OFFS(p)	((p)->a)
+#define MEMREF_SIZE(p)		((p)->b)
+
+struct tee_supp_msg {
+	uint32_t cmd_ret;
+	uint32_t num_param;
+	struct tee_param params[TEE_REQ_PARAM_MAX];
+};
+
+struct param_value {
+	uint64_t a;
+	uint64_t b;
+	uint64_t c;
+};
+
+static int receive_request(const struct device *dev, struct tee_supp_msg *ts_req)
+{
+	int rc = tee_suppl_recv(dev, &ts_req->cmd_ret, &ts_req->num_param, ts_req->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant receive failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static int send_response(const struct device *dev, struct tee_supp_msg *rsp)
+{
+	int rc = tee_suppl_send(dev, rsp->cmd_ret, rsp->num_param, rsp->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant send response failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static void uuid_from_octets(TEEC_UUID *d, const uint8_t s[TEE_UUID_LEN])
+{
+	d->timeLow = (s[0] << 24) | (s[1] << 16) | (s[2] << 8) | s[3];
+	d->timeMid = (s[4] << 8) | s[5];
+	d->timeHiAndVersion = (s[6] << 8) | s[7];
+	memcpy(d->clockSeqAndNode, s + 8, sizeof(d->clockSeqAndNode));
+}
+
+static int load_ta(uint32_t num_params, struct tee_param *params)
+{
+	int ta_found = 0;
+	size_t size = 0;
+	uint8_t val_cmd[TEE_UUID_LEN];
+	TEEC_UUID uuid = { 0 };
+	struct tee_shm *shm;
+	void *addr;
+
+	if (num_params != 2) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	if ((params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) != TEE_PARAM_ATTR_TYPE_VALUE_INPUT ||
+	    (params[1].attr & TEE_PARAM_ATTR_TYPE_MASK) != TEE_PARAM_ATTR_TYPE_MEMREF_OUTPUT) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	((uint64_t *)val_cmd)[0] = params[0].a;
```

Comment:
This is the easiest way to catch an alignment fault  exception. `val_cmd` being byte array can be aligned to any address, while `uint64_t` value needs to be aligned to 8 bytes. 

Also this violates MISRA C Rule 11.3:

> A cast shall not be performed between a pointer to object type and a pointer to a different object type.

Correct and portable  way to convert `uint64_t` into bunch of `uint8_t`s is bit operations:

```
val_cmd[0] = params[0].a & 0xFF;
val_cmd[1] = params[0].a << 8 & 0xFF; 
val_cmd[2] = params[0].a << 16 & 0xFF; 
[... and so on...]
```
Just mind your endianness

Reply:
I guess `memcpy` will be better

Reply:
Tricky question? What endianness ARM uses?

Reply:
You guess endianess will be different in OP-TEE and real world? This value is not multi-byte value. This is byte array placed in two uint64_t fields.

Reply:
It can be. You can take SCMI as reference. This protocol also passes multiple config values in a couple of uint64_t and it pays attention to endianess.

Reply:
Document GPD_TEE_Internal_Core_API_Specification_v1.3.1_PublicRelease_CC has the following note:

> Note: The TEE_UUID structure is sensitive to differences in the endianness of the Client API and the TA.
> It is the responsibility of the Trusted OS to ensure that any endianness difference between client and TA is
> managed internally when those structures are passed through one of the defined APIs. The definition below
> assumes that the endianness of both Client API and TA are the same, and needs to be changed
> appropriately if this is not the case.

in Section 3.2.4 page 45 which says that it is Trusted OS responsibility to handle endianess between TA and Client API and UUID definition assumes that TA and Client API has the same endianess.

Reply:
Oh, okay. Thank you for the research